doc/compiler/control/CheckpointRestoreCoordination.md

```diff
@@ -0,0 +1,434 @@
+   Monitor
+5. The Checkpointing Thread wakes up with the CR Monitor in hand, releases the
+   CR Monitor, and blocks on the Comp Monitor
+6. The Comp Threads release the Comp Monitor and wait on the CompThread Monitor
+7. The Checkpointing Thread acquires the Comp Monitor, ensures the state has
+   changed to `COMPTHREAD_SUSPENDED` for the current `compInfoPT`
+8. The Checkpointing Thread checks the next `compInfoPT`, waiting on the CR
+   Monitor if needed (it will release the Comp Monitor prior to waiting)
+9. The Checkpointing Thread proceeds with the remaining coordination tasks
+
+In steps 7 and 8, in addition to checking the Compilation Thread State, the
+Checkpointing Thread also checks if the checkpoint should be interrupted, and
+returns from `TR::CompilationInfo::suspendCompThreadsForCheckpoint` if so. This
+is because when the Checkpointing Thread releases the Comp Monitor to wait on
+the CR Monitor, the Shutdown Thread can acquire the Comp Monitor to begin the
+shutdown process.
+
+### JIT Dump
+
+#### Crash on compilation thread
+* The crashing Comp Thread will
+[return to the VM and terminate the process](https://github.com/eclipse-openj9/openj9/blob/500e0a26e2c5be6ead0f838495ae8d8cc34821e1/runtime/compiler/control/CompilationThread.cpp#L3442-L3447).
+* Possible scenarios for Checkpointing Thread:
+    1. The Checkpointing Thread will try to acquire the Comp Monitor to signal
+       the Comp Threads to suspend themselves but will end up blocking until the
+       JVM terminates.
+    2. The Checkpointing Thread will manage to signal Comp Threads to suspend,
+       but will remain waiting on the CR Monitor for the crashed Comp Thread to
+       suspend until the JVM terminates.
+
+#### Crash on application thread
+This scenario should not be possible, as in normal checkpoint, all application
+threads will have already been suspended by this point. However, for the sake of
+completeness, this scenario does not differ from the process described in the
+[Normal Execution](#normal-execution-steps-5--6-can-be-interchanged) section.
+
+### Normal Shutdown
+
+#### Scenario 1
+1. The Checkpointing Thread waits on CR Monitor
+2. The Comp Thread goes to suspend and already has Comp Monitor in hand
+3. The Shutdown Thread blocks on the Comp Monitor
+4. The Comp Thread updates its state to `COMPTHREAD_SUSPENDED`, acquires the CR
+   Monitor, and calls `notifyAll`
+5. The Comp Thread releases the CR Monitor, releases the Comp Monitor, and waits
+   on CompThread Monitor
+6. The Checkpointing Thread wakes up with the CR Monitor in hand, releases the
+   CR Monitor, acquires the Comp Monitor and continues on to next thread
+7. The Checkpointing Thread checks the state of the Comp Thread and moves onto
+   the next threads possibly finding them all suspended.
+8. The Checkpointing Thread releases the Comp Monitor and returns from the hook
+9. The Shutdown Thread goes through the sequence of stopping all threads
```

Comment:
Would it be better to explicitly mention in step 9 that the Shutdown Thread goes through the sequence of stopping all threads after the Comp Monitor is released by the Checkpointing Thread that has suspended all compilation threads?

Reply:
I could, though the Shutdown Thread doesn't necessarily hold on the Comp Monitor for the entire process; it only does so for stopping the Compilation Threads. I think the fact that this happens in the order specified (and there is no comment saying the order can be changed) should suffice.